File: 01-游戏开发/04-游戏引擎/02-Archetype/go/archetype.go
// archetype.go — archetype（原型）存储的最小实现（与 python/archetype.py 同题）。
//
// Unity Entities 官方定义：archetype = 相同组件类型组合的唯一标识；chunk = 16KiB
// 均匀内存块，块内每组件一个数组 + 一个 entity id 数组，紧密排布；删除时把块内
// 最后一个实体搬来填空；块满则新建、块空则销毁。
package main

import (
	"fmt"
	"sort"
)

const chunkSize = 16 * 1024 // 16 KiB
const entityIDSize = 4      // demo 约定

var componentSize = map[string]int{"Position": 8, "Velocity": 8, "Health": 4}

// keyOf 把组件名集合规范成 archetype 的键（排序后拼接，与顺序无关）。
func keyOf(names []string) string {
	sorted := append([]string(nil), names...)
	sort.Strings(sorted)
	out := ""
	for i, n := range sorted {
		if i > 0 {
			out += "+"
		}
		out += n
	}
	return out
}

func memberList(key string) []string {
	if key == "" {
		return []string{}
	}
	out := []string{}
	cur := ""
	for i := 0; i < len(key); i++ {
		if key[i] == '+' {
			out = append(out, cur)
			cur = ""
			continue
		}
		cur += string(key[i])
	}
	return append(out, cur)
}

func entitySize(key string) int {
	total := entityIDSize
	for _, c := range memberList(key) {
		total += componentSize[c]
	}
	return total
}

func chunkCapacity(key string) int { return chunkSize / entitySize(key) }

type chunk struct {
	key      string
	capacity int
	arrays   map[string][]float64
	ids      []int
	count    int
}

func newChunk(key string) *chunk {
	cap := chunkCapacity(key)
	c := &chunk{key: key, capacity: cap, ids: make([]int, cap), count: 0}
	c.arrays = map[string][]float64{}
	for _, name := range memberList(key) {
		c.arrays[name] = make([]float64, cap)
	}
	return c
}

type archetype struct {
	key    string
	chunks []*chunk
}

func (a *archetype) freeChunk() *chunk {
	for _, c := range a.chunks {
		if c.count < c.capacity {
			return c
		}
	}
	c := newChunk(a.key)
	a.chunks = append(a.chunks, c)
	return c
}

type location struct {
	key   string
	chunk *chunk
	index int
}

type world struct {
	archetypes map[string]*archetype
	loc        map[int]*location
	comps      map[int]map[string]float64
	nextID     int
	copies     int
}

func newWorld() *world {
	return &world{
		archetypes: map[string]*archetype{},
		loc:        map[int]*location{},
		comps:      map[int]map[string]float64{},
	}
}

func (w *world) archetypeFor(key string) *archetype {
	a, ok := w.archetypes[key]
	if !ok {
		a = &archetype{key: key}
		w.archetypes[key] = a
	}
	return a
}

func (w *world) place(e int, a *archetype) {
	c := a.freeChunk()
	idx := c.count
	c.ids[idx] = e
	for name, arr := range c.arrays {
		arr[idx] = w.comps[e][name]
		w.copies++
	}
	c.count++
	w.loc[e] = &location{key: a.key, chunk: c, index: idx}
}

func (w *world) unplace(e int) {
	loc := w.loc[e]
	delete(w.loc, e)
	c := loc.chunk
	last := c.count - 1
	moved := c.ids[last]
	if moved != e { // swap-remove：块内最后一个实体填空
		c.ids[loc.index] = moved
		for name, arr := range c.arrays {
			arr[loc.index] = arr[last]
		}
		w.loc[moved] = &location{key: c.key, chunk: c, index: loc.index}
	}
	c.count--
}

func (w *world) create(comps map[string]float64) int {
	e := w.nextID
	w.nextID++
	names := []string{}
	for n := range comps {
		names = append(names, n)
	}
	w.comps[e] = comps
	a := w.archetypeFor(keyOf(names))
	w.place(e, a)
	return e
}

func (w *world) destroy(e int) {
	loc := w.loc[e]
	w.unplace(e)
	delete(w.comps, e)
	if loc.chunk.count == 0 {
		a := w.archetypes[loc.key]
		for i, c := range a.chunks {
			if c == loc.chunk {
				a.chunks = append(a.chunks[:i], a.chunks[i+1:]...)
				break
			}
		}
	}
}

// setComponent 覆盖组件值；name 为新组件时等价于 add 的结构变更。
func (w *world) setComponent(e int, name string, value float64) {
	w.comps[e][name] = value
	w.unplace(e)
	names := []string{}
	for n := range w.comps[e] {
		names = append(names, n)
	}
	w.place(e, w.archetypeFor(keyOf(names)))
}

func (w *world) removeComponent(e int, name string) {
	delete(w.comps[e], name)
	w.unplace(e)
	names := []string{}
	for n := range w.comps[e] {
		names = append(names, n)
	}
	w.place(e, w.archetypeFor(keyOf(names)))
}

func (w *world) query(want ...string) []int {
	out := []int{}
	for key, a := range w.archetypes {
		if !superset(key, want) {
			continue
		}
		for _, c := range a.chunks {
			out = append(out, c.ids[:c.count]...)
		}
	}
	return out
}

func superset(key string, want []string) bool {
	have := map[string]bool{}
	for _, n := range memberList(key) {
		have[n] = true
	}
	for _, n := range want {
		if !have[n] {
			return false
		}
	}
	return true
}

func main() {
	capPV := chunkCapacity(keyOf([]string{"Position", "Velocity"}))
	fmt.Printf("chunk 容量(Position+Velocity) = %d (16KiB / %d B)\n", capPV,
		entitySize(keyOf([]string{"Position", "Velocity"})))

	w := newWorld()
	ids := []int{}
	for i := 0; i < capPV+1; i++ {
		ids = append(ids, w.create(map[string]float64{"Position": float64(i), "Velocity": 1}))
	}
	a := w.archetypes[keyOf([]string{"Position", "Velocity"})]
	fmt.Printf("chunk 数 = %d, 首块占用 = %d/%d\n", len(a.chunks), a.chunks[0].count, a.chunks[0].capacity)

	filler := a.chunks[0].ids[a.chunks[0].count-1]
	w.destroy(a.chunks[0].ids[0])
	fmt.Printf("swap-remove 后 ids[0] = %d (filler=%d), count=%d, loc[ filler ].index=%d\n",
		a.chunks[0].ids[0], filler, a.chunks[0].count, w.loc[filler].index)

	w2 := newWorld()
	e := w2.create(map[string]float64{"Position": 0, "Velocity": 2})
	before := w2.copies
	w2.setComponent(e, "Health", 10)
	fmt.Printf("加组件搬移 %d 个组件值, 新 archetype = %s\n", w2.copies-before, w2.loc[e].key)
	before = w2.copies
	w2.removeComponent(e, "Health")
	fmt.Printf("删组件搬移 %d 个组件值, 回到 archetype = %s\n", w2.copies-before, w2.loc[e].key)

	w3 := newWorld()
	x := w3.create(map[string]float64{"Position": 1})
	y := w3.create(map[string]float64{"Position": 2, "Velocity": 2})
	w3.create(map[string]float64{"Health": 3})
	got := w3.query("Position")
	fmt.Printf("query(Position) = %v (期望含 %d,%d)\n", got, x, y)
	fmt.Printf("query(Position,Velocity) = %v\n", w3.query("Position", "Velocity"))
}

File: 01-游戏开发/04-游戏引擎/02-Archetype/python/archetype.py
"""archetype.py — archetype（原型）存储的最小实现与自检.

Unity Entities 官方定义（docs.unity3d.com/Packages/com.unity.entities@1.0/manual/concepts-archetypes.html）：

  * archetype 是「同一世界中拥有相同组件类型组合的实体」的唯一标识；
    增删组件会把实体搬到对应的 archetype，不存在则新建。
  * 同一 archetype 的实体与组件存放在 **chunk** 里，每个 chunk **16KiB**，
    能放多少实体取决于该 archetype 组件的数量与大小。
  * chunk 内每个组件类型一个数组，外加一个存 entity id 的数组；数组**紧密排布**，
    新实体放第一个空位，删除时把 **chunk 内最后一个实体搬来填空**。
  * 所有 chunk 满 → 新建 chunk；chunk 内最后一个实体被移走 → 销毁 chunk。

本 demo 用 Python 复刻这套布局，并把「增删组件 = 跨 archetype 搬移全部组件」的代价量化出来。
"""

from __future__ import annotations

from typing import Dict, FrozenSet, Iterable, List, Optional, Tuple

CHUNK_SIZE = 16 * 1024            # 16 KiB —— Unity 文档原话
ENTITY_ID_SIZE = 4                # demo 约定：entity id 占 4 字节（非 Unity 官方数值）
COMPONENT_SIZE = {"Position": 8, "Velocity": 8, "Health": 4}

Key = FrozenSet[str]

_ASSERTIONS = 0


def check(label: str, cond: bool, detail: str = "") -> None:
    global _ASSERTIONS
    if not cond:
        raise AssertionError(f"{label} 失败: {detail}")
    _ASSERTIONS += 1
    print(f"ok {_ASSERTIONS:>2} {label}" + (f"  [{detail}]" if detail else ""))


def key_of(names: Iterable[str]) -> Key:
    return frozenset(names)


def entity_size(key: Key) -> int:
    return ENTITY_ID_SIZE + sum(COMPONENT_SIZE[c] for c in key)


def chunk_capacity(key: Key) -> int:
    return CHUNK_SIZE // entity_size(key)


class Chunk:
    """一个 16KiB 的均匀内存块：每组件一个数组 + 一个 entity id 数组。"""

    def __init__(self, key: Key) -> None:
        self.key = key
        self.capacity = chunk_capacity(key)
        self.arrays: Dict[str, List[Optional[float]]] = {
            c: [None] * self.capacity for c in sorted(key)
        }
        self.ids: List[Optional[int]] = [None] * self.capacity
        self.count = 0

    def __repr__(self) -> str:  # pragma: no cover - 仅调试
        return f"<Chunk {sorted(self.key)} {self.count}/{self.capacity}>"


class Archetype:
    def __init__(self, key: Key) -> None:
        self.key = key
        self.chunks: List[Chunk] = []

    def free_chunk(self) -> Chunk:
        for ch in self.chunks:
            if ch.count < ch.capacity:
                return ch
        ch = Chunk(self.key)
        self.chunks.append(ch)
        return ch


class World:
    def __init__(self) -> None:
        self.archetypes: Dict[Key, Archetype] = {}
        self.loc: Dict[int, Tuple[Key, Chunk, int]] = {}
        self.comps: Dict[int, Dict[str, float]] = {}
        self.next_id = 0
        self.copies = 0            # 结构变更中「写入 chunk 的组件个数」累计
        self.arch_version = 0      # 新建 archetype 时 +1（查询缓存据此失效）

    # ---- 内部 -------------------------------------------------------
    def _archetype(self, key: Key) -> Archetype:
        arch = self.archetypes.get(key)
        if arch is None:
            arch = Archetype(key)
            self.archetypes[key] = arch
            self.arch_version += 1
        return arch

    def _place(self, eid: int, arch: Archetype) -> None:
        ch = arch.free_chunk()
        idx = ch.count
        ch.ids[idx] = eid
        for c, arr in ch.arrays.items():
            arr[idx] = self.comps[eid][c]
            self.copies += 1
        ch.count += 1
        self.loc[eid] = (arch.key, ch, idx)

    def _unplace(self, eid: int) -> None:
        _key, ch, idx = self.loc.pop(eid)
        last = ch.count - 1
        moved = ch.ids[last]
        if moved != eid:                       # swap-remove：chunk 内最后一个实体填空
            ch.ids[idx] = moved
            for c, arr in ch.arrays.items():
                arr[idx] = arr[last]
            self.loc[moved] = (ch.key, ch, idx)
        ch.ids[last] = None
        for arr in ch.arrays.values():
            arr[last] = None
        ch.count -= 1

    # ---- 公开 API ---------------------------------------------------
    def create(self, **comps: float) -> int:
        eid = self.next_id
        self.next_id += 1
        self.comps[eid] = dict(comps)
        self._place(eid, self._archetype(key_of(comps)))
        return eid

    def destroy(self, eid: int) -> None:
        _key, ch, _idx = self.loc[eid]
        self._unplace(eid)
        self.comps.pop(eid, None)
        if ch.count == 0:
            arch = self.archetypes[ch.key]
            arch.chunks.remove(ch)

    def add(self, eid: int, name: str, value: float) -> None:
        self.comps[eid][name] = value
        self._unplace(eid)
        self._place(eid, self._archetype(key_of(self.comps[eid])))

    def remove(self, eid: int, name: str) -> None:
        self.comps[eid].pop(name)
        self._unplace(eid)
        self._place(eid, self._archetype(key_of(self.comps[eid])))

    def archetype_of(self, eid: int) -> Key:
        return self.loc[eid][0]

    def query(self, *names: str) -> List[int]:
        want = set(names)
        out: List[int] = []
        for key, arch in self.archetypes.items():
            if not want <= set(key):
                continue
            for ch in arch.chunks:
                for i in range(ch.count):
                    out.append(ch.ids[i])          # type: ignore[arg-type]
        return out


class Query:
    """缓存匹配 archetype 列表；新建 archetype（arch_version 变化）时重算。"""

    def __init__(self, world: World, *names: str) -> None:
        self.world = world
        self.want = set(names)
        self.version = -1
        self.recomputes = 0
        self.matched: List[Archetype] = []

    def archetypes(self) -> List[Archetype]:
        if self.version != self.world.arch_version:
            self.matched = [a for k, a in self.world.archetypes.items() if self.want <= set(k)]
            self.version = self.world.arch_version
            self.recomputes += 1
        return self.matched

    def run(self) -> List[int]:
        out: List[int] = []
        for arch in self.archetypes():
            for ch in arch.chunks:
                out.extend(ch.ids[: ch.count])     # type: ignore[misc]
        return out


def main() -> None:
    cap_pv = chunk_capacity(key_of(["Position", "Velocity"]))
    check("chunk 容量 = 16KiB / 单实体字节数", cap_pv == CHUNK_SIZE // (4 + 8 + 8),
          f"capacity={cap_pv}")
    check("空 archetype 容量最大", chunk_capacity(key_of([])) == CHUNK_SIZE // ENTITY_ID_SIZE,
          f"{chunk_capacity(key_of([]))}")
    check("组件越多容量越小", chunk_capacity(key_of(["Position", "Velocity", "Health"])) < cap_pv)

    w = World()
    ids = [w.create(Position=float(i), Velocity=1.0) for i in range(cap_pv + 1)]
    arch_pv = w.archetypes[key_of(["Position", "Velocity"])]
    check("超过容量才新建 chunk", len(arch_pv.chunks) == 2, f"chunks={len(arch_pv.chunks)}")
    check("首个 chunk 是满的", arch_pv.chunks[0].count == cap_pv, f"{arch_pv.chunks[0].count}")
    check("chunk 内紧密排布（id 连续）", arch_pv.chunks[0].ids[:3] == ids[:3],
          f"{arch_pv.chunks[0].ids[:3]}")
    check("组件数组与 id 数组同下标对齐",
          arch_pv.chunks[0].arrays["Position"][5] == 5.0 and arch_pv.chunks[0].ids[5] == ids[5])

    # swap-remove 只在 chunk 内部发生（不是跨 chunk 搬）
    first_chunk_ids_before = list(arch_pv.chunks[0].ids[: arch_pv.chunks[0].count])
    victim = arch_pv.chunks[0].ids[0]
    filler = first_chunk_ids_before[-1]
    w.destroy(victim)
    check("删除后由本 chunk 最后一个实体填空", arch_pv.chunks[0].ids[0] == filler,
          f"ids[0]={arch_pv.chunks[0].ids[0]} filler={filler}")
    check("被搬移者的 loc 已回写", w.loc[filler][2] == 0, f"idx={w.loc[filler][2]}")
    check("被删实体已无定位", victim not in w.loc)
    check("count 递减、数组无空洞",
          arch_pv.chunks[0].count == cap_pv - 1
          and all(arch_pv.chunks[0].ids[i] is not None for i in range(arch_pv.chunks[0].count)))

    # chunk 空了就销毁
    w2 = World()
    solo = w2.create(Position=1.0)
    w2.destroy(solo)
    check("chunk 内最后一个实体被移走即销毁 chunk",
          len(w2.archetypes[key_of(["Position"])].chunks) == 0)

    # 结构变更：加/删组件 = 换 archetype 并搬走全部组件
    w3 = World()
    e = w3.create(Position=0.0, Velocity=2.0)
    before = w3.copies
    w3.add(e, "Health", 10.0)
    check("加组件后换了 archetype", w3.archetype_of(e) == key_of(["Position", "Velocity", "Health"]))
    check("旧组件数据保留", w3.comps[e]["Position"] == 0.0 and w3.comps[e]["Velocity"] == 2.0)
    check("搬移代价 = 新组合的组件数", w3.copies - before == 3, f"copies={w3.copies - before}")
    key3 = key_of(["Position", "Velocity", "Health"])
    check("组件值在新 chunk 里已落盘",
          w3.archetypes[key3].chunks[0].arrays["Health"][0] == 10.0)
    before = w3.copies
    w3.remove(e, "Health")
    check("删组件回到原 archetype", w3.archetype_of(e) == key_of(["Position", "Velocity"]))
    check("删组件同样按新组合搬移", w3.copies - before == 2, f"copies={w3.copies - before}")

    # archetype 唯一性 / 稳定性
    w4 = World()
    w4.create(Position=1.0, Velocity=1.0)
    w4.create(Velocity=1.0, Position=2.0)
    check("组件顺序不影响 archetype", len(w4.archetypes) == 1)
    n_arch = len(w4.archetypes)
    for i in range(50):
        w4.create(Position=float(i), Velocity=1.0)
    check("archetype 集合很快稳定", len(w4.archetypes) == n_arch, f"{len(w4.archetypes)}")

    # 查询：按「archetype 组件集合是查询的超集」匹配
    w5 = World()
    a = w5.create(Position=1.0)
    b = w5.create(Position=2.0, Velocity=2.0)
    w5.create(Health=3.0)
    check("查询单组件命中所有超集", sorted(w5.query("Position")) == sorted([a, b]),
          f"{w5.query('Position')}")
    check("查询两组件只命中同时拥有者", w5.query("Position", "Velocity") == [b])
    check("查询不到的组合返回空", w5.query("Position", "Health") == [])

    q = Query(w5, "Position")
    q.run()
    r1 = q.recomputes
    w5.create(Position=9.0)                      # 已有 archetype，不触发重算
    q.run()
    check("已有 archetype 不使查询缓存失效", q.recomputes == r1, f"{q.recomputes}")
    w5.create(Position=8.0, Health=1.0)          # 新 archetype → 缓存失效
    q.run()
    check("新建 archetype 使查询缓存失效", q.recomputes == r1 + 1, f"{q.recomputes}")
    check("缓存查询结果与全量扫描一致", sorted(q.run()) == sorted(w5.query("Position")))

    # 迭代连续性：查询输出按 chunk 紧密排列
    arch = w5.archetypes[key_of(["Position"])]
    check("迭代输出无空洞",
          all(arch.chunks[0].ids[i] is not None for i in range(arch.chunks[0].count)))

    print(f"\n全部 {_ASSERTIONS} 条断言通过")


if __name__ == "__main__":
    main()
